### DISCUSSION TITLE
"changed_at" column does not get updated on password update when "password_expiration" is enabled

### DISCUSSION THREAD

**Discussion by joseargao** [jeremyevans/rodauth#427] [General]
We are looking into potentially using rodauth for our app's authentication needs, and one of our required features is password expiration. Our problem is that the "changed_at" column in our DB does not get updated after using the change_password feature. We have set the following in our rodauth_main.rb:

```
enable :create_account, :verify_account, :verify_account_grace_period,
           :login, :logout, :jwt, :jwt_refresh,
           :reset_password, :change_password, :change_password_notify,
           :change_login, :verify_login_change, :close_account,
           :password_expiration

    password_expiration_id_column :account_id
    password_expiration_changed_at_column :changed_at
    require_password_change_after 30.days
```

We also added a table to our db:

```
class CreateAccountPasswordChangeTimes < ActiveRecord::Migration[7.1]
  def change
    create_table :account_password_change_times do |t|
      t.integer :account_id, null: false
      t.timestamp :changed_at, null: false, default: -> { 'CURRENT_TIMESTAMP' }
    end

    add_index :account_password_change_times, :account_id, unique: true
    add_foreign_key :account_password_change_times, :accounts
  end
end
```

Everything seems to work as expected, except the "changed_at" column does not get updated automatically upon password change. We worked around it with this:

```
    after_change_password do
      user_id = account_from_session[:id]
      unless user_id.nil?
        password_change_record = AccountPasswordChangeTime.find_by(account_id: user_id)
        password_change_record&.update(changed_at: Time.now)
      end
    end
```

However, it feels like this is not the right way to get this feature working.

**Comment by jeremyevans** [comment]
Something seems wrong.  The `change_password` feature calls `set_password`, and `password_expiration` overrides that to update the password change time.

If you can put together a failing spec or minimum self contained example showing the issue (using only plain Rodauth and not rodauth-rails), I would be happy to research this issue.

**Reply by joseargao** [reply]
Thank you for the response @jeremyevans. I have been playing around with a self-contained example, but I have not been able to observe the same behavior using only plain rodauth. Instead I ran into a different scenario that I also found to be curious.

I was checking the changed_at value in the DB before and after change-password, sleeping for one second in between. While rodauth was updating the value as expected it was being set to an entirely different timezone. The output looks like:

```
changed_at before change-password 2024-06-18 06:15:41 +0900
...
changed_at after change-password 2024-06-18 15:15:43 +0900
```

I also tried to configure my DB in the spec_helper like so:

```
config.before(:suite) do
    ...
    DB.timezone = :utc
    Sequel.application_timezone = :local
    ...
end
```

I tried to set various values above but I could never get a consistent timezone for the before and after. Am I missing a step? I put up the project I used for my testing here: https://github.com/joseargao/rodauth_change_password

**Reply by jeremyevans** [reply]
Looks like you are using SQLite, which doesn't have a native timestamp type.  If you want to store timestamps with timezones on SQLite, you probably need to set `DB.use_timestamp_timezones = true`.  Note that if you do that, SQLite's date/time functions will not work correctly with the timestamps.

If you are storing the database timezone in UTC, you may want to set `DB.current_timestamp_utc = true` (Sequel converts it to localtime by default, but that may interact poorly with `DB.timezone = :utc`).

I would try `DB.current_timestamp_utc = true` first and see if that fixes things.  A better fix would be using a database with an actual timestamp type, such as PostgreSQL.

**Reply by joseargao** [reply]
Thank you for the response once again. This makes sense to me now. We will definitely change up the DB for the actual app. This was just me trying to replicate the strange behavior of the changed_at column in a minimal sample.

If I ever figure out what was up with that behavior, I will post here again.

**Reply by janko** [reply]
I tried running your spec with Sequel using Active Record's DB connection instead of its own, to replicate the conditions you most likely have in your Rails app, but was getting the same result. Setting `DB.current_timestamp_utc = true` resolves the timezone differences in the "after" of `changed_at`, so now they're only 1-2 seconds apart for me (`sequel-activerecord_connection` sets that automatically).

**Reply by joseargao** [reply]
@janko thank you for looking into this. Just to clarify things, the issue I encountered in my Rails test app is different from the timezone issue I ran into in my Ruby test app. Basically, the "changed_at" column doesn't automatically get updated in the Rails version and I don't know why.

I actually put the [Rails test app](https://github.com/joseargao/rodauth_rails_change_password) on github as well, in case you or anyone else would be interested in taking a look and letting me know what's up. It's not a dealbreaker problem because we can still manually update this column in the DB. It's just strange that it doesn't get done automatically (and only in the Rails version).

**Reply by janko** [reply]
@joseargao I found the issue in your Rails app. You're relying on Timecop making Rodauth set `changed_at` to the travelled time, but Rodauth updates the column using `CURRENT_TIMESTAMP`, which retrieves the current time at the database side. It doesn't generate a time object in Ruby, so it's unaffected by Timecop.

If you turn on Active Record logging around the `POST /change-password` request in your tests, you can clearly see the `changed_at` being updated:

```
Sequel (0.3ms)  UPDATE "account_password_change_times" SET "changed_at" = CURRENT_TIMESTAMP WHERE ("account_id" = 1)
Sequel (0.3ms)  UPDATE "accounts" SET "password_hash" = '$2a$04$IQkcYPlm1mkY6PjrvjcnSurTaF4m0OrBfn6F6jgn9/.oDLN7edcfW' WHERE ("id" = 1)
```

**Reply by joseargao** [reply]
@janko thanks for pointing me in the right direction! It does indeed look like the DB is being updated as expected based on the Active Record logging. Perhaps Timecop was the wrong tool for me to test this.

**Reply by joseargao** [reply]
In case anyone wanders into this thread in the future, I have updated my rails project so it can be used as a reference for how I resolved my issues. https://github.com/joseargao/rodauth_rails_change_password
